// vue-whiteboard-canvas/src/containers/whiteboard/hooks/listeners/tool-listeners.js
import useDrawPencil from "../tools/useDrawPencil";
import {useWhiteboardStore} from "../../store/whiteboard";
import useDrawLine from "../tools/useDrawLine";
import useText from "../tools/useText";
import usePan from "../tools/usePan";
import useMove from "../tools/useMove";
import useDrawRect from "../tools/useDrawRect";
import useDrawCircle from "../tools/useDrawCircle";

const {stage, layer, states} = useWhiteboardStore();
const {onPencilMouseMove, onPencilMouseUp, onPencilMouseDown} = useDrawPencil(stage, layer);
const {onRectMouseMove, onRectMouseUp, onRectMouseDown} = useDrawRect(stage, layer);
const {onCircleMouseMove, onCircleMouseUp, onCircleMouseDown} = useDrawCircle(stage, layer);
const {onLineMouseMove, onLineMouseDown, onLineMouseUp} = useDrawLine(stage, layer);
const {onTextMouseMove, onTextMouseDown, onTextMouseUp} = useText(stage, layer);
const {onPanMouseMove, onPanMouseUp, onPanMouseDown} = usePan();
const {onMoveMouseMove, onMoveMouseUp, onMoveMouseDown} = useMove();


export const listeners = {
  pencil: {
    onMouseDown: onPencilMouseDown,
    onMouseUp: onPencilMouseUp,
    onMouseMove: onPencilMouseMove
  },
  line: {
    onMouseDown: onLineMouseDown,
    onMouseUp: onLineMouseUp,
    onMouseMove: onLineMouseMove,
  },
  move: {
    onMouseDown: onMoveMouseDown,
    onMouseUp: onMoveMouseUp,
    onMouseMove: onMoveMouseMove,
  },
  text: {
    onMouseDown: onTextMouseDown,
    onMouseUp: onTextMouseUp,
    onMouseMove: onTextMouseMove,
  },
  pan: {
    onMouseDown: onPanMouseDown,
    onMouseUp: onPanMouseUp,
    onMouseMove: onPanMouseMove,
  },
  rectangle: {
    onMouseDown: onRectMouseDown,
    onMouseUp: onRectMouseUp,
    onMouseMove: onRectMouseMove,
  },
  circle: {
    onMouseDown: onCircleMouseDown,
    onMouseUp: onCircleMouseUp,
    onMouseMove: onCircleMouseMove,
  }

}
