// vue-whiteboard-canvas/src/containers/whiteboard/utils/delta-store.js
import localforage from "localforage";
import { v4 as uuidV4 } from 'uuid';
import {useHistoryStore} from "../store/history";
import {useWhiteboardStore} from "../store/whiteboard";

const deltaStore = localforage.createInstance({
  driver      : localforage.INDEXEDDB, // Force WebSQL; same as using setDriver()
  name        : '__vue_whiteboard',
  version     : 0.1,
  size        : 4980736, // Size of database, in bytes. WebSQL-only for now.
  storeName   : '__vue_whiteboard_quill_deltas', // Should be alphanumeric, with underscores.
  description : 'store quill deltas to convert into canvas images by app.'
});

const saveDelta = async function (delta) {
  const id = uuidV4();
  await deltaStore.setItem(id, delta);
  return id;
}

const loadDelta = function (deltaId) {
  return deltaStore.getItem(deltaId);
}

// clear everything in store
const clearAll = function (){
  return deltaStore.clear()
}

/** remove everything else except:
 * 1. delta associated with undo history
 * 2. delta associated in all quill images in layer
 * */
const pruneDelta = function (){
  const {history} = useHistoryStore();
  const {layer} = useWhiteboardStore();

  // get an array of delta id from history
  // loop thru history
  // for each history array, extract the delta ids

  // get an array of delta ids from layer

  // filter out all deltas in history


}

export {
  deltaStore,
  saveDelta,
  loadDelta,
  clearAll,
};
